// oidc-client-node/lib/oidc-client.js
'use strict';

/**
 * Module dependencies
 */

var promise = require("bluebird");
var r = require('jsrsasign');
var utility = require('./utility.js');
var Cookies = require('cookies');
var url = require('url');

function OidcClient(req, res, serverSettings) {

  if (!(this instanceof OidcClient)) { return new OidcClient(req, res, serverSettings); }

  var self = this;
    
  // if you don't clone, you end up keeping the request / response state on the server level configuration object.
  self._settings = utility.copy(serverSettings);

  utility.verbose_logging = self._settings.verbose_logging || false;

  if (!self._settings.request_state_key) {
    self._settings.request_state_key = "OidcClient.request_state";
  }

  if (!self._settings.request_state_store) {
    self._settings.request_state_store = new Cookies(req, res);
  }

  if (typeof self._settings.load_user_profile === 'undefined') {
    self._settings.load_user_profile = true;
  }

  if (typeof self._settings.filter_protocol_claims === 'undefined') {
    self._settings.filter_protocol_claims = true;
  }

  if (self._settings.authority && self._settings.authority.indexOf('.well-known/openid-configuration') < 0) {
    if (self._settings.authority[self._settings.authority.length - 1] !== '/') {
      self._settings.authority += '/';
    }
    self._settings.authority += '.well-known/openid-configuration';
  }

  if (!self._settings.response_type) {
    self._settings.response_type = "id_token token";
  }

  if (self._settings.post_logout_redirect_uri) {
    self._settings.post_logout_redirect_uri = verifyFullyQualifiedUrl(self._settings.post_logout_redirect_uri, req);
  }

  if (self._settings.redirect_uri) {
    self._settings.redirect_uri = verifyFullyQualifiedUrl(self._settings.redirect_uri, req);
  }
}

function generateError(message) {
  return promise.reject(Error(message));
}

function responseTypeRegEx(self, expectedResult) {
  if (self._settings.response_type) {
    var result = self._settings.response_type.split(/\s+/g).filter(function (item) {
      return item === expectedResult;
    });
    return !!(result[0]);
  }
  return false;
}

function originalURL(req) {
  var headers = req.headers;
  var protocol = (req.secure || req.headers['x-forwarded-proto'] === 'https') ? 'https' : 'http';
  var host = headers.host;
  var path = req.url || '';

  // Check for azure header
  if (!req.secure && req.headers["x-arr-ssl"]) {
    protocol = 'https';
  }

  return protocol + '://' + host + path;
}

function verifyFullyQualifiedUrl(callbackUrl, req) {
  if (callbackUrl) {
    var parsed = url.parse(callbackUrl);
    if (!parsed.protocol) {
      // The callback URL is relative, resolve a fully qualified URL from the
      // URL of the originating request.
      callbackUrl = url.resolve(originalURL(req), callbackUrl);
    }
  }
  return callbackUrl;
}

OidcClient.prototype = {

  isOidc: function () {
    var self = this;
    return responseTypeRegEx(self, "id_token");
  },

  isOauth: function () {
    var self = this;
    return responseTypeRegEx(self, "token");
  },

  isCode: function () {
    var self = this;
    return responseTypeRegEx(self, "code");
  },

  loadMetadataAsync: function () {
    utility.log("OidcClient.loadMetadataAsync");
    var self = this;
    var settings = self._settings;

    if (settings.metadata) {
      return promise.resolve(settings.metadata);
    }

    if (!settings.authority) {
      return generateError("No authority configured");
    }

    return utility.getJson(settings.authority, null, settings.httpSettings)
      .then(function (metadata) {
        utility.log("OidcClient.loadMetadataAsync.then");
        settings.metadata = metadata;
        return metadata;
      }).catch(function (error) {
        utility.log("OidcClient.loadMetadataAsync.error");
        var errorMessage = "Failed to load metadata: ";
        if (error && error.message) {
          errorMessage = errorMessage.concat(error.message);
        }
        return generateError(errorMessage);
      });
  },

  loadX509SigningKeyAsync: function () {
    utility.log("OidcClient.loadX509SigningKeyAsync");
    var self = this;
    var settings = self._settings;

    function getKeyAsync(jwks) {
      if (!jwks.keys || !jwks.keys.length) {
        return generateError("Signing keys empty");
      }

      var key = jwks.keys[0];
      if (key.kty !== "RSA") {
        return generateError("Signing key not RSA");
      }

      if (!key.x5c || !key.x5c.length) {
        return generateError("RSA keys empty");
      }

      return promise.resolve(key.x5c[0]);
    }

    if (settings.jwks) {
      return getKeyAsync(settings.jwks);
    }

    return self.loadMetadataAsync().then(function (metadata) {
      if (!metadata.jwks_uri) {
        return generateError("Metadata does not contain jwks_uri");
      }

      return utility.getJson(metadata.jwks_uri, null, settings.httpSettings).then(function (jwks) {
        settings.jwks = jwks;
        return getKeyAsync(jwks);
      }).catch(function (err) {
        return generateError("Failed to load signing keys (" + err.message + ")");
      });
    });
  },

  loadUserProfile: function (access_token) {
    utility.log("OidcClient.loadUserProfile");
    var self = this;

    return self.loadMetadataAsync().then(function (metadata) {

      if (!metadata.userinfo_endpoint) {
        return promise.reject(Error("Metadata does not contain userinfo_endpoint"));
      }

      return utility.getJson(metadata.userinfo_endpoint, access_token, self._settings.httpSettings);
    });
  },

  loadAuthorizationEndpoint: function () {
    utility.log("OidcClient.loadAuthorizationEndpoint");
    var self = this;

    if (self._settings.authorization_endpoint) {
      return promise.resolve(self._settings.authorization_endpoint);
    }

    if (!self._settings.authority) {
      return generateError("No authorization_endpoint configured");
    }

    return self.loadMetadataAsync().then(function (metadata) {
      utility.log("OidcClient.loadAuthorizationEndpoint.then");
      if (!metadata || !metadata.authorization_endpoint) {
        utility.log("OidcClient.loadAuthorizationEndpoint.returnError");
        var errorMessage = "Metadata does not contain authorization_endpoint";
        return generateError(errorMessage);
      }

      return metadata.authorization_endpoint;
    });
  },
  
    loadTokenEndpoint: function () {
    utility.log("OidcClient.loadTokenEndpoint");
    var self = this;

    if (self._settings.token_endpoint) {
      return promise.resolve(self._settings.token_endpoint);
    }

    if (!self._settings.authority) {
      return generateError("No token_endpoint configured");
    }

    return self.loadMetadataAsync().then(function (metadata) {
      utility.log("OidcClient.loadTokenEndpoint.then");
      if (!metadata || !metadata.token_endpoint) {
        utility.log("OidcClient.loadTokenEndpoint.returnError");
        var errorMessage = "Metadata does not contain token_endpoint";
        return generateError(errorMessage);
      }

      return metadata.token_endpoint;
    });
  },

  createTokenRequestAsync: function (user_state) {
    utility.log("OidcClient.createTokenRequestAsync");

    var self = this;
    var settings = self._settings;

    return self.loadAuthorizationEndpoint().then(function (authorization_endpoint) {
      utility.log("OidcClient.createTokenRequestAsync.then");
      var state = utility.rand();
      var url = authorization_endpoint + "?state=" + encodeURIComponent(state);

      var nonce = null;

      if (self.isOidc()) {
        nonce = utility.rand();
        url += "&nonce=" + encodeURIComponent(nonce);
      }

      var required = ["client_id", "redirect_uri", "response_type", "scope"];
      required.forEach(function (key) {
        var value = settings[key];
        if (value) {
          url += "&" + key + "=" + encodeURIComponent(value);
        }
      });

      var optional = ["prompt", "display", "max_age", "ui_locales", "id_token_hint", "login_hint", "acr_values", "response_mode"];
      optional.forEach(function (key) {
        var value = settings[key];
        if (value) {
          url += "&" + key + "=" + encodeURIComponent(value);
        }
      });

      var request_state = {
        oidc: self.isOidc(),
        oauth: self.isOauth(),
        code: self.isCode(),
        state: state,
        user_state: user_state
      };

      if (nonce) {
        request_state.nonce = nonce;
      }

      settings.request_state_store.set(settings.request_state_key, JSON.stringify(request_state));

      return {
        request_state: request_state,
        url: url
      };
    });
  },

  createLogoutRequestAsync: function (id_token_hint) {
    utility.log("OidcClient.createLogoutRequestAsync");
    var self = this;
    var settings = self._settings;

    return self.loadMetadataAsync().then(function (metadata) {
      if (!metadata.end_session_endpoint) {
        return generateError("No end_session_endpoint in metadata");
      }

      var url = metadata.end_session_endpoint;
      if (id_token_hint && settings.post_logout_redirect_uri) {
        url += "?post_logout_redirect_uri=" + encodeURIComponent(settings.post_logout_redirect_uri);
        url += "&id_token_hint=" + encodeURIComponent(id_token_hint);
      }
      return url;
    });
  },

  validateIdTokenAsync: function (id_token, nonce, access_token) {
    utility.log("OidcClient.validateIdTokenAsync");

    var self = this;
    var settings = self._settings;

    return self.loadX509SigningKeyAsync().then(function (cert) {

      var jws = new r.jws.JWS();
      if (jws.verifyJWSByPemX509Cert(id_token, cert)) {
        var id_token_contents = JSON.parse(jws.parsedJWS.payloadS);

        if (nonce !== id_token_contents.nonce) {
          return generateError("Invalid nonce");
        }

        return self.loadMetadataAsync().then(function (metadata) {
          if (id_token_contents.iss !== metadata.issuer) {
            return generateError("Invalid issuer");
          }

          if (id_token_contents.aud !== settings.client_id) {
            return generateError("Invalid audience");
          }

          var now = Math.round(Date.now() / 1000);

          // accept tokens issues up to 5 mins ago
          var diff = now - id_token_contents.iat;
          if (diff > (5 * 60)) {
            return generateError("Token issued too long ago");
          }

          if (id_token_contents.exp < now) {
            return generateError("Token expired");
          }

          if (access_token && settings.load_user_profile) {
            // if we have an access token, then call user info endpoint
            return self.loadUserProfile(access_token, id_token_contents).then(function (profile) {
              if (profile.statusCode) {
                // profile is an http response, not a profile response
                return utility.copy(id_token_contents);
              } else {
                return utility.copy(profile, id_token_contents);
              }

            });
          } else {
              // no access token, so we have all our claims
              return id_token_contents;
            }
        });
      } else {
          return generateError("JWT failed to validate");
        }
    });
  },

  validateAccessTokenAsync: function (id_token_contents, access_token) {
    utility.log("OidcClient.validateAccessTokenAsync");

    if (!id_token_contents.at_hash) {
      return generateError("No at_hash in id_token");
    }

    var hash = r.crypto.Util.sha256(access_token);
    var left = hash.substr(0, hash.length / 2);
    var left_b64u = r.hextob64u(left);

    if (left_b64u !== id_token_contents.at_hash) {
      return generateError("at_hash failed to validate");
    }

    return promise.resolve();
  },

  validateIdTokenAndAccessTokenAsync: function (id_token, nonce, access_token) {
    utility.log("OidcClient.validateIdTokenAndAccessTokenAsync");

    var self = this;

    return self.validateIdTokenAsync(id_token, nonce, access_token).then(function (id_token_contents) {

      return self.validateAccessTokenAsync(id_token_contents, access_token).then(function () {

        return id_token_contents;

      });

    });
  },
  
  getTokenFromCodeAsync: function (code) {
    utility.log("OidcClient.getTokenFromCodeAsync");
    var self = this;
    var settings = self._settings;
    
    return self.loadTokenEndpoint().then(function (token_endpoint) {
      var config = {
        method: 'POST',
        form: {
          grant_type: 'authorization_code',
          code: code,
          redirect_uri: settings.redirect_uri,
          client_id: settings.client_id,
          client_secret: settings.client_secret
        }
      };
      
      utility.copy(self._settings.httpSettings, config);
      
      return utility.getJson(token_endpoint, null, config);
    });
  },

  getRequestState: function (){
      var self = this;
      var settings = self._settings;
    
      var request_state = settings.request_state_store.get(settings.request_state_key);  
      if (request_state){
        return JSON.parse(request_state);
      }else{
          return {};
      }
  },
  
  processResponseAsync: function (result, requestState) {
    utility.log("OidcClient.processResponseAsync");

    var self = this;
    var settings = self._settings;

    var request_state = requestState;

    if (!request_state) {
      request_state = settings.request_state_store.get(settings.request_state_key);
      settings.request_state_store.set(settings.request_state_key);
    }

    if (!request_state) {
      return generateError("No request state loaded");
    }

    request_state = JSON.parse(request_state);
    if (!request_state) {
      return generateError("No request state loaded");
    }

    if (!request_state.state) {
      return generateError("No state loaded");
    }

    if (!result) {
      return generateError("No OIDC response");
    }

    if (result.error) {
      return generateError(result.error);
    }

    if (result.state !== request_state.state) {
      return generateError("Invalid state");
    }

    if (request_state.oidc) {
      if (!result.id_token) {
        return generateError("No identity token");
      }

      if (!request_state.nonce) {
        return generateError("No nonce loaded");
      }
    }

    if (request_state.oauth) {
      if (!result.access_token) {
        return generateError("No access token");
      }

      if (!result.token_type || result.token_type.toLowerCase() !== "bearer") {
        return generateError("Invalid token type");
      }

      if (!result.expires_in) {
        return generateError("No token expiration");
      }
    }

    if (request_state.code) {
      if (!result.code) {
        return generateError("No auth code");
      }
    }

    var localPromise = promise.resolve();
    if (request_state.oidc && request_state.oauth) {
      localPromise = self.validateIdTokenAndAccessTokenAsync(result.id_token, request_state.nonce, result.access_token);
    } else if (request_state.oidc) {
      localPromise = self.validateIdTokenAsync(result.id_token, request_state.nonce);
    } else if (request_state.code) {
      localPromise = self.getTokenFromCodeAsync(result.code).then(function (codeResult) {
        result = codeResult;
        return self.validateIdTokenAsync(result.id_token, request_state.nonce);
      })
    }

    return localPromise.then(function (profile) {
      if (profile && settings.filter_protocol_claims) {
        var remove = ["nonce", "at_hash", "iat", "nbf", "exp", "aud", "iss", "idp"];
        remove.forEach(function (key) {
          delete profile[key];
        });
      }
      
      return {
        profile: profile,
        id_token: result.id_token,
        access_token: result.access_token,
        refresh_token: result.refresh_token,
        expires_in: result.expires_in,
        scope: result.scope,
        session_state: result.session_state,
        user_state: request_state.user_state
      };
    });
  },

  mergeRequestOptions: function (req, options) {

    var self = this;

    // if we get extra properties here that aren't allowed, the actual request will strip them out with the property whitelist
    utility.copy(options, self._settings);
    
    var callbackURL = options.callbackURL || self._settings.callbackURL;
    self._settings.redirect_uri = verifyFullyQualifiedUrl(callbackURL, req);

    var logoutCallbackUrl = options.post_logout_redirect_uri || self._settings.post_logout_redirect_uri;
    self._settings.post_logout_redirect_uri = verifyFullyQualifiedUrl(logoutCallbackUrl, req);   

    var scope = options.scope || self._settings.scope;
    if (Array.isArray(scope)) { scope = scope.join(self._settings.scopeSeparator); }
    if (scope) {
      self._settings.scope = 'openid' + self._settings.scopeSeparator + scope;
    } else {
      self._settings.scope = 'openid';
    }

    utility.log(self._settings);
  },
  
  refreshAccessTokenAsync: function (refreshToken) {
    utility.log("OidcClient.refreshAccessToken");
    var self = this;
    var settings = self._settings;
    
    return self.loadTokenEndpoint().then(function (token_endpoint) {
      var config = {
      method: 'POST',
      form: {
        grant_type: 'refresh_token',
        refresh_token: refreshToken,
        redirect_uri: settings.redirect_uri,
        client_id: settings.client_id,
        client_secret: settings.client_secret
        }
      };
      
      utility.copy(self._settings.httpSettings, config);
      
      return utility.getJson(token_endpoint, null, config).then(function (token) {
        return token;
      });
    });
  }
};

module.exports = OidcClient;
